### DISCUSSION TITLE
Golang: Change HTTP ClientRequest type to DataFlow::CallNode

### DISCUSSION THREAD

**Discussion by jason-invision** [github/securitylab#184] [General] [answered]
Could the type of ClientRequest be changed to a DataFlow::CallNode?

In the javascript version of CodeQL the type is an InvokeNode, which allows for 
```
      exists(ClientRequest req | sink = req.getAnArgument())
```

In the golang version, I had to do 
```
    exists(HTTP::ClientRequest request, DataFlow::CallNode call | request = call and sink = call.getAnArgument())
```

**Comment by max-schaefer** [comment, marked as answer]
I'm not sure. I think keeping the more general type has its advantage, just in case we end up wanting to model an API where the client request isn't actually a call. Note that you can use postfix casts to suppress most of the syntactic noise:

```ql
exists(HTTP::ClientRequest req | sink = req.(DataFlow::CallNode).getAnArgument())
```

Longer term, I think it would be worth considering introducing a member predicate on `HTTP::ClientRequest` that does what you are using `getAnArgument()` for here, though I'm not entirely sure what that is.
